pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -0,0 +1,160 @@
+  // Considered a Gemm operation.
+  // For now they are affine.load, affined.load, mul(f/i), reduce add, terminator.
+  const unsigned kNumValidInstrInGemmRegion = 5;
+
+ public:
+  explicit Stencil() {};
+  
+  // Main function
+  void DoStenciling(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext);
+
+  // Is this a Gemm operation
+  mlir::Optional<GemmOperationMatch> getGemmOperation(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext);
+};
+
+mlir::Optional<GemmOperationMatch> Stencil::getGemmOperation(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext) {
+  mlir::Optional<GemmOperationMatch> ret;
+  Builder builder(mlirContext);
+  auto *body = op.getBody();
+  // Get the instructions in the body and match for load, load, mulXXX, reduce add operations.
+  // For everything else we fail.
+  if (body->getOperations().size() != kNumValidInstrInGemmRegion) {
+    op.emitRemark("the ParallelOp region didn't have the right number of instructions for a Gemm");
+    IVLOG(3, "the ParallelOp region didn't have the right number of instructions for a Gemm");
+    return ret;
+  }
+
+  op.walk([&](AffineReduceOp reduceOp) {
+    IVLOG(3, "Found ReduceOp");
+
+    // Not check the reduceOp aggregation.
+    if (reduceOp.agg() != AggregationKind::add) {
+      op.emitRemark("the reduce operation is not addition");
+      IVLOG(3, "the reduce operation is not addition");
+      return;
+    }
+
+    // Get the in tensors for the reduce op.
+    Value reduceIn = reduceOp.val();
+    MulOperationType mulOpType = MulOperationType::NoneMulOpType;
+
+    // Make sure the in for the reduce is a result of a multiplication.
+    auto valDef = reduceIn.getDefiningOp();
+
+    if (valDef == nullptr) {
+      op.emitRemark("the source of the reduce operation is not defined in this block");
+      IVLOG(3, "the source of the reduce operation is not defined in this block");
+      return;
+    }
+
+    mlir::MulFOp mulfOp  = llvm::dyn_cast_or_null<mlir::MulFOp>(valDef);
+    mlir::MulIOp muliOp  = llvm::dyn_cast_or_null<mlir::MulIOp>(valDef);
+    if (mulfOp == nullptr && muliOp == nullptr) {
+      op.emitRemark("the source of the reduce is not a multiplication operation");
+      IVLOG(3, "The source of the reduce is not a multiplication operation");
+      return;
+    }
+
+    mlir::AffineLoadOp lhs;
+    mlir::AffineLoadOp rhs;
+    if (mulfOp != nullptr) {
+      mulOpType = MulOperationType::FloatTy;
+      lhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(mulfOp.lhs().getDefiningOp());
+      rhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(mulfOp.rhs().getDefiningOp());
+    }
+    else if (muliOp != nullptr) {
+      mulOpType = MulOperationType::IntTy;
+      lhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(muliOp.lhs().getDefiningOp());
+      rhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(muliOp.rhs().getDefiningOp());
+    }
+    else {
+      op.emitRemark("unhandled multiplication type in stenciling");
```

Comment:
Isn't this code unreachable given the early check on mulfOp == nullptr && muliOp == nullptr?

Reply:
You are right... In general, I like to have reminders in code when it is likely to add more cases (like BF6 mul op for example at some point). And have the run/code tell me what is wrong instead of spending time trying to figure out what went wrong. Took it out, though...